File: Cpp/VersionHistoryDialog.cpp
#include "VersionHistoryDialog.h"
#include "Session.h"

#include <QDateTime>
#include <QHBoxLayout>
#include <QListWidget>
#include <QListWidgetItem>
#include <QPlainTextEdit>
#include <QPushButton>
#include <QVBoxLayout>

namespace {

QString formatStamp(const QString &stamp)
{
    const QDateTime dt = QDateTime::fromString(stamp, "yyyyMMddTHHmmsszzz");
    return dt.isValid() ? dt.toString("dd/MM/yyyy HH:mm:ss") : stamp;
}

} // namespace

VersionHistoryDialog::VersionHistoryDialog(const QString &draftId, QWidget *parent)
    : QDialog(parent)
    , m_draftId(draftId)
{
    setWindowTitle("Historique des versions");
    resize(600, 420);

    m_listWidget = new QListWidget(this);
    connect(m_listWidget, &QListWidget::currentItemChanged, this, &VersionHistoryDialog::showPreview);

    m_preview = new QPlainTextEdit(this);
    m_preview->setReadOnly(true);

    m_restoreBtn = new QPushButton("Restaurer cette version", this);
    auto *closeBtn = new QPushButton("Fermer", this);
    connect(m_restoreBtn, &QPushButton::clicked, this, &VersionHistoryDialog::restoreSelected);
    connect(closeBtn, &QPushButton::clicked, this, &VersionHistoryDialog::close);

    auto *contentRow = new QHBoxLayout();
    contentRow->addWidget(m_listWidget, 1);
    contentRow->addWidget(m_preview, 2);

    auto *buttonsRow = new QHBoxLayout();
    buttonsRow->addStretch();
    buttonsRow->addWidget(m_restoreBtn);
    buttonsRow->addWidget(closeBtn);

    auto *layout = new QVBoxLayout(this);
    layout->addLayout(contentRow);
    layout->addLayout(buttonsRow);

    for (const QString &stamp : Session::listVersions(draftId)) {
        auto *item = new QListWidgetItem(formatStamp(stamp));
        item->setData(Qt::UserRole, stamp);
        m_listWidget->addItem(item);
    }
    if (m_listWidget->count() > 0)
        m_listWidget->setCurrentRow(0);
}

void VersionHistoryDialog::showPreview(QListWidgetItem *current)
{
    if (!current) {
        m_preview->setPlainText(QString());
        return;
    }
    m_preview->setPlainText(Session::readVersion(m_draftId, current->data(Qt::UserRole).toString()));
}

void VersionHistoryDialog::restoreSelected()
{
    QListWidgetItem *item = m_listWidget->currentItem();
    if (!item)
        return;
    emit restoreRequested(Session::readVersion(m_draftId, item->data(Qt::UserRole).toString()));
    close();
}

File: Python/version_history_dialog.py
from datetime import datetime

from PyQt6.QtCore import Qt, pyqtSignal
from PyQt6.QtWidgets import (
    QDialog,
    QHBoxLayout,
    QListWidget,
    QListWidgetItem,
    QPlainTextEdit,
    QPushButton,
    QVBoxLayout,
)

import session


def _format_stamp(stamp):
    try:
        return datetime.strptime(stamp, "%Y%m%dT%H%M%S%f").strftime("%d/%m/%Y %H:%M:%S")
    except ValueError:
        return stamp


class VersionHistoryDialog(QDialog):
    restore_requested = pyqtSignal(str)

    def __init__(self, draft_id, parent=None):
        super().__init__(parent)
        self.draft_id = draft_id
        self.setWindowTitle("Historique des versions")
        self.resize(600, 420)

        self.list_widget = QListWidget(self)
        self.list_widget.currentItemChanged.connect(self._show_preview)

        self.preview = QPlainTextEdit(self)
        self.preview.setReadOnly(True)

        self.restore_btn = QPushButton("Restaurer cette version", self)
        self.close_btn = QPushButton("Fermer", self)
        self.restore_btn.clicked.connect(self._restore_selected)
        self.close_btn.clicked.connect(self.close)

        content_row = QHBoxLayout()
        content_row.addWidget(self.list_widget, 1)
        content_row.addWidget(self.preview, 2)

        buttons_row = QHBoxLayout()
        buttons_row.addStretch()
        buttons_row.addWidget(self.restore_btn)
        buttons_row.addWidget(self.close_btn)

        layout = QVBoxLayout(self)
        layout.addLayout(content_row)
        layout.addLayout(buttons_row)

        for stamp in session.list_versions(draft_id):
            item = QListWidgetItem(_format_stamp(stamp))
            item.setData(Qt.ItemDataRole.UserRole, stamp)
            self.list_widget.addItem(item)
        if self.list_widget.count() > 0:
            self.list_widget.setCurrentRow(0)

    def _show_preview(self, current, _previous):
        if current is None:
            self.preview.setPlainText("")
            return
        stamp = current.data(Qt.ItemDataRole.UserRole)
        self.preview.setPlainText(session.read_version(self.draft_id, stamp))

    def _restore_selected(self):
        item = self.list_widget.currentItem()
        if item is None:
            return
        stamp = item.data(Qt.ItemDataRole.UserRole)
        self.restore_requested.emit(session.read_version(self.draft_id, stamp))
        self.close()
